### PR TITLE
improve r_buf_buffer and remove r_buf_get_at

### PR DESCRIPTION
**Pull request by ret2libc**
r2r PR: https://github.com/radare/radare2-regressions/pull/1786

### CODE REVIEW

File: libr/bin/p/bin_mbn.c [hunk 1]
```diff
@@ -77,14 +77,14 @@ static bool load_bytes(RBinFile *bf, void **bin_obj, const ut8 *buf, ut64 sz, ut
 
 static bool load(RBinFile *bf) {
 	if (bf && bf->buf) {
-		const ut8 *bytes = r_buf_buffer (bf->buf);
-		ut64 sz = r_buf_size (bf->buf);
+		ut64 sz;
+		const ut8 *bytes = r_buf_buffer (bf->buf, &sz);
```
Comment (radare) [inline libr/bin/p/bin_mbn.c]: i think r_buf_buffer is a really bad name

**Comment by radare** [comment]
do you know that a buffer can be used as a frontend for io, and this can be 64bit in size? this is a bunch of terabytes.. i doubt you want to do that

**Comment by radare** [comment]
yeah i cant think of a better name either

**Comment by ret2libc** [comment]
> do you know that a buffer can be used as a frontend for io, and this can be 64bit in size? this is a bunch of terabytes.. i doubt you want to do that

I know. That's why you shouldn't use `r_buf_buffer` in the first place :) It's inherently wrong.

**Comment by radare** [comment]
Fhen how do you read data without copying it entirely? :p when u know the data is there

**Comment by ret2libc** [comment]
How do you know "the data is there"? You may know it only if you created the RBuffer in that same function, but... at that point there's no need to use RBuffer if you just want the entire buffer.

In all other cases, you can't be sure that "the data is there". The whole point of RBuffer is that you need to read small chunks when you need them, because the data may not be there.

**Comment by radare** [comment]
> On 9 Apr 2019, at 21:49, Riccardo Schirone <notifications@github.com> wrote:
> 
> How do you know "the data is there"? You may know it only if you created the RBuffer in that same function, but... at that point there's no need to use RBuffer if you just want the entire buffer.
> 
> 

because if its a malloc, the data is there. any other option that is not malloc or mmap it shouldnt allow to access that raw buffer pointer+fixedsize

**Comment by radare** [comment]
yes, reading small chunks, but not copying it entirely, for example a fatmacho file can contain 10 binaries of 10M each, and u only want to load one, theres no need to copy all that memory several times to finally get few bytes to check the header. the copy of a bfufer must be an explicit operation. which can be done in a constructor for example.

**Comment by radare** [comment]
and instead of r_buf_buffer() (which maybe should be named "r_buf_weakdata()" or so.. there must be an r_buf_clone() oor r_buf_dup() to duplicate the buffer contents

**Comment by ret2libc** [comment]
> because if its a malloc, the data is there. any other option that is not malloc or mmap it shouldnt allow to access that raw buffer pointer+fixedsize

My point is, if you have this:
```C
void myfunction(RBuffer *b, int a) {
    const ut8 *t = r_buf_buffer (b);
    ...
}
```

How can myfunction know that it will always receive a malloc/mmap-based RBuffer? It can't. So, either you make r_buf_buffer work with all kinds of RBuffers, or you can't use it at all. That said, you shouldn't use r_buf_buffer in any case, but you should always assume the data needs to be loaded from somewhere. If you don't want to do that, then I don't know why we are using Rbuffer.

**Comment by radare** [comment]
Thats the thing:

* we dont want to use this api
* the api only makes sense on malloc cases where the user is in control of what it have created
* copying data is an unnecessary step because if u read previsous points its clear that its not going to be necessRy

**Comment by ret2libc** [comment]
> * we dont want to use this api

I agree

> * the api only makes sense on malloc cases where the user is in control of what it have created

I partially agree. The point is: RBuffer can be of different types. The power of using RBuffer is that you have an API that behaves the same indipendently of how it was created. So if you have a RBuffer backed by IO, you can do the same operations as if you have a RBuffer backed by malloc or one backed by file or mmap.

In the moment you add some operations that are valid only for some type of RBuffers, you are defeating the purpose of RBuffer, and it means when the RBuffer will switch his type, nothing will work anymore.

And almost never you can be sure of what's the type of RBuffer. For example, the `RBinPlugin.load` method does accept a RBuffer, but you can't say whether it comes from malloc, mmap, io or whatever else.

> * copying data is an unnecessary step because if u read previsous points its clear that its not going to be necessRy

Please provide example with code. Try to create a function that accepts a RBuffer and needs to compare 10 bytes at offset 0x200 with an hard-coded string. How would you do that?

File: libr/bin/p/bin_mbn.c (hunk 1 above)
Comment (ret2libc) [reply libr/bin/p/bin_mbn.c]: `r_buf_data`?

**Comment by ret2libc** [comment]
@radare I think this is ready to be reviewed. Somehow travis is still red, but I think it's because of master... as soon as master gets green, i can make sure this is green as well.

**Comment by radare** [comment]
ok to merge

**Pull request merged**